maraboupy/MarabouNetworkNNet.py

```diff
@@ -263,31 +563,346 @@ def buildEquations(self):
         for layer, size in enumerate(hidden_layers):
             for node in range(size):
-                self.addRelu(self.nodeTo_b(layer+1, node), self.nodeTo_f(layer+1, node))
+                self.addRelu(self.nodeTo_b(layer + 1, node), self.nodeTo_f(layer + 1, node))
 
     def numberOfVariables(self):
-        return self.layerSizes[0] + 2*sum(self.layerSizes[1:-1]) + 1*self.layerSizes[-1]
+        return self.layerSizes[0] + 2 * sum(self.layerSizes[1:-1]) + 1 * self.layerSizes[-1]
 
     def getInputMinimum(self, inputVar):
         return self.inputMinimums[inputVar]
 
     def getInputMaximum(self, inputVar):
         return self.inputMaximums[inputVar]
+
+
+    def evaluateNetwork(self, inputs, normalize_inputs=True, normalize_outputs=True, activate_output_layer=False):
+
+        """
+        Evaluate the network directly, without Marabou
+
+        Args:
+            inputs (list of floats): Network inputs to be evaluated
+
+        Returns:
+            (numpy array of floats): Network output
+       """
+
+        numLayers = self.numLayers
+        inputSize = self.inputSize
+        outputSize = self.outputSize
+        biases = self.biases
+        weights = self.weights
+        mins = self.inputMinimums
+        maxes = self.inputMaximums
+        means = self.inputMeans
+        ranges = self.inputRanges
+
+        # Prepare the inputs to the neural network
+        if (normalize_inputs):
+            inputsNorm = np.zeros(inputSize)
+            for i in range(inputSize):
+                if inputs[i] < mins[i]:
+                    inputsNorm[i] = (mins[i] - means[i]) / ranges[i]
+                elif inputs[i] > maxes[i]:
+                    inputsNorm[i] = (maxes[i] - means[i]) / ranges[i]
+                else:
+                    inputsNorm[i] = (inputs[i] - means[i]) / ranges[i]
+        else:
+            inputsNorm = inputs
+
+        # Evaluate the neural network
+        for layer in range(numLayers - 1):
+            inputsNorm = np.maximum(np.dot(weights[layer], inputsNorm) + biases[layer], 0)
+
+        if (activate_output_layer):
+            outputs = np.maximum(np.dot(weights[-1], inputsNorm) + biases[-1], 0)
+        else:
+            outputs = np.dot(weights[-1], inputsNorm) + biases[-1]
+
+        # Undo output normalization
+        if (normalize_outputs):
+            for i in range(outputSize):
+                outputs[i] = outputs[i] * self.outputRange + self.outputMean
+
+        return outputs
+
+    def evaluateNetworkMultiple(self, inputs, normalize_inputs=True, normalize_outputs=True,
+                                activate_output_layer=False):
+
+        """
+         Evaluate network using multiple sets of inputs
+
+         Args:
+             inputs (numpy array of lists of floats): Array of network inputs to be evaluated.
+
+         Returns:
+             (numpy array of numpy arrays floats): Network outputs for each set of inputs
+
+        HAS NOT BEEN TESTED
```

Comment:
Please either test or remove this function

Reply:
Also, why not invoke the previous functions multiple times? This seems like code duplication

Reply:
The evaluate...Multiple functions are not necessary. These functions seem to be mostly copied from the NNet repository, and code for evaluating multiple points at once was implemented in the NNet repository to improve computation time when evaluating a large number of points. For Marabou, I don't think there is a need for this multiple-evaluation function since we just want to check a few points to show that Marabou equations produce the same results as an NNet evaluation.

Reply:
Actually, as a part of computing statistics for hidden neurons, as well as for checking a candidate for an invariant (without invoking Marabou), I do evaluate the networks on tens of thousands of inputs at the time. So I thought that I eventually I would use evaluate...Multiple for this, precisely for improving efficiency. At the moment I have removed them from this PR. When and if I start using them, I can add them back (together with appropriate tests).